nifi-nar-bundles/nifi-hbase-bundle/nifi-hbase-processors/src/main/java/org/apache/nifi/hbase/ScanHBase.java

```diff
@@ -0,0 +1,564 @@
+                hBaseClientService.scan(tableName,
+                                        startRow, endRow,
+                                        filterExpression,
+                                        timerangeMin, timerangeMax,
+                                        limitRows,
+                                        isReversed,
+                                        columns,
+                                        handler);
+            } catch (IOException e) {
+                getLogger().error("Unable to fetch rows from HBase table {} due to {}", new Object[] {tableName, e});
+                flowFile = session.putAttribute(flowFile, "scanhbase.results.found", Boolean.toString(handler.isHandledAny()));
+                session.transfer(flowFile, REL_FAILURE);
+                return;
+            }
+
+            LinkedList<Tuple<byte[], ResultCell[]>> hangingRows = handler.getHangingRows();
+            if (!handler.isHandledAny() ||                                                         // no rows found in hbase
+                    (handler.isHandledAny() && (hangingRows == null || hangingRows.isEmpty()))     // all the rows are flushed to FF inside handlers
+                ){
+                flowFile = session.putAttribute(flowFile, "scanhbase.results.found", Boolean.toString(handler.isHandledAny()));
+                session.transfer(flowFile, REL_ORIGINAL);
+                session.commit();
+                return;
+            }
+
+            if (hangingRows != null && !hangingRows.isEmpty()) {
+                FlowFile lastFF = session.create(flowFile);
+                final Map<String, String> attributes = new HashMap<>();
+                attributes.put(HBASE_TABLE_ATTR, tableName);
+                attributes.put(HBASE_ROWS_COUNT_ATTR, Long.toString(rowsPulledHolder.get()));
+                attributes.put(CoreAttributes.MIME_TYPE.key(), "application/json");
+                attributes.put(HBASE_ROWS_COUNT_ATTR, Long.toString(hangingRows.size()));
+                lastFF = session.putAllAttributes(lastFF, attributes);
+
+                final AtomicReference<IOException> ioe = new AtomicReference<>(null);
+                lastFF = session.write(lastFF, (out) -> {
+                    for (Iterator<Tuple<byte[], ResultCell[]>> iter = hangingRows.iterator(); iter.hasNext();){
+                        Tuple<byte[], ResultCell[]> r = iter.next();
+                        serializer.serialize(r.getKey(), r.getValue(), out);
+                        if (iter.hasNext()){
+                            out.write(nl);
+                        }
+                    }
+                });
+
+                Relationship rel = REL_SUCCESS;
+                IOException error = ioe.get();
+                if (error != null){
+                    lastFF = session.putAttribute(lastFF, "scanhbase.error", error.toString());
+                    rel = REL_FAILURE;
+                }
+                session.transfer(lastFF, rel);
```

Comment:
Looking at the code later, it seems like lines 398-423 are the same as 513-539, except there is no generation of a provenance event in this part and there is in the later part.

Assuming we do want a provenance event here and they are the same, then it might make sense to separate this code out into a private method that is called from both places.